File: src/benchmark/metrics.js
export function scoreFilePredictions({ predictedFiles, acceptedFiles }) {
  const accepted = new Set(acceptedFiles);
  const predicted = [...new Set(predictedFiles)];
  const hits = predicted.filter((file) => accepted.has(file));
  const precisionAt5 = fraction(hits.filter((file) => predicted.indexOf(file) < 5).length, Math.min(5, predicted.length));
  const recallAt5 = fraction(acceptedFiles.filter((file) => predicted.slice(0, 5).includes(file)).length, accepted.size);
  const recallAt10 = fraction(acceptedFiles.filter((file) => predicted.slice(0, 10).includes(file)).length, accepted.size);
  const firstRank = firstAcceptedRank(predicted, accepted);

  return {
    acceptedFiles,
    predictedFiles: predicted,
    hits,
    hitAt1: firstRank === 1,
    hitAt5: firstRank > 0 && firstRank <= 5,
    recallAt5: Number(recallAt5.toFixed(3)),
    recallAt10: Number(recallAt10.toFixed(3)),
    precisionAt5: Number(precisionAt5.toFixed(3)),
    mrr: firstRank > 0 ? Number((1 / firstRank).toFixed(3)) : 0,
    firstAcceptedRank: firstRank || null
  };
}

export function aggregateScores(results) {
  const completed = results.filter((result) => result.status === "ok");
  return {
    total: results.length,
    completed: completed.length,
    failed: results.length - completed.length,
    avgRecallAt5: average(completed.map((result) => result.metrics.recallAt5)),
    avgRecallAt10: average(completed.map((result) => result.metrics.recallAt10)),
    avgPrecisionAt5: average(completed.map((result) => result.metrics.precisionAt5)),
    avgMrr: average(completed.map((result) => result.metrics.mrr)),
    hitAt1: completed.filter((result) => result.metrics.hitAt1).length,
    hitAt5: completed.filter((result) => result.metrics.hitAt5).length
  };
}

function firstAcceptedRank(predicted, accepted) {
  for (let index = 0; index < predicted.length; index += 1) {
    if (accepted.has(predicted[index])) {
      return index + 1;
    }
  }
  return 0;
}

function average(values) {
  if (!values.length) {
    return 0;
  }
  return Number((values.reduce((sum, value) => sum + value, 0) / values.length).toFixed(3));
}

function fraction(numerator, denominator) {
  if (!denominator) {
    return 0;
  }
  return numerator / denominator;
}

File: go_issue_agent/benchmark/metrics.py
def score_file_predictions(predicted_files: list[str], accepted_files: list[str]) -> dict:
    accepted = set(accepted_files)
    predicted = list(dict.fromkeys(predicted_files))
    hits = [file for file in predicted if file in accepted]
    precision_at_5 = _fraction(len([file for file in hits if predicted.index(file) < 5]), min(5, len(predicted)))
    recall_at_5 = _fraction(len([file for file in accepted_files if file in predicted[:5]]), len(accepted))
    recall_at_10 = _fraction(len([file for file in accepted_files if file in predicted[:10]]), len(accepted))
    first_rank = _first_accepted_rank(predicted, accepted)
    return {
        "acceptedFiles": accepted_files,
        "predictedFiles": predicted,
        "hits": hits,
        "hitAt1": first_rank == 1,
        "hitAt5": first_rank > 0 and first_rank <= 5,
        "recallAt5": round(recall_at_5, 3),
        "recallAt10": round(recall_at_10, 3),
        "precisionAt5": round(precision_at_5, 3),
        "mrr": round(1 / first_rank, 3) if first_rank > 0 else 0,
        "firstAcceptedRank": first_rank or None,
    }


def aggregate_scores(results: list[dict]) -> dict:
    completed = [result for result in results if result.get("status") == "ok"]
    return {
        "total": len(results),
        "completed": len(completed),
        "failed": len(results) - len(completed),
        "avgRecallAt5": _average([result["metrics"]["recallAt5"] for result in completed]),
        "avgRecallAt10": _average([result["metrics"]["recallAt10"] for result in completed]),
        "avgPrecisionAt5": _average([result["metrics"]["precisionAt5"] for result in completed]),
        "avgMrr": _average([result["metrics"]["mrr"] for result in completed]),
        "hitAt1": len([result for result in completed if result["metrics"]["hitAt1"]]),
        "hitAt5": len([result for result in completed if result["metrics"]["hitAt5"]]),
    }


def _first_accepted_rank(predicted: list[str], accepted: set[str]) -> int:
    for index, value in enumerate(predicted, start=1):
        if value in accepted:
            return index
    return 0


def _average(values: list[float]) -> float:
    return round(sum(values) / len(values), 3) if values else 0


def _fraction(numerator: int, denominator: int) -> float:
    return numerator / denominator if denominator else 0
